File: platforms/android/source/TokamakPlugins.kt
package com.tokamak.runtime

import android.app.Activity
import android.net.Uri
import android.webkit.WebView
import androidx.webkit.JavaScriptReplyProxy
import androidx.webkit.WebMessageCompat
import androidx.webkit.WebViewCompat
import org.json.JSONObject

internal class TokamakPluginError(
    val errorName: String,
    message: String,
) : Exception(message) {
    companion object {
        fun notSupported(message: String) = TokamakPluginError("NotSupportedError", message)
    }
}

internal typealias TokamakPluginReply = (Result<Any?>) -> Unit

internal interface TokamakPlugin {
    val id: String

    fun call(method: String, arguments: Any?, reply: TokamakPluginReply) {
        reply(Result.failure(TokamakPluginError.notSupported("$id.$method is not supported")))
    }

    fun subscribe(
        method: String,
        arguments: Any?,
        reply: TokamakPluginReply,
    ): () -> Unit {
        reply(Result.failure(TokamakPluginError.notSupported("$id.$method is not supported")))
        return {}
    }

    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ) = Unit
}

internal class TokamakPluginBridge(
    private val activity: Activity,
    private val host: String,
    plugins: List<TokamakPlugin>,
) : WebViewCompat.WebMessageListener {
    private data class RequestKey(val session: String, val id: Int)

    private val plugins =
        buildMap {
            plugins.forEach { plugin ->
                check(put(plugin.id, plugin) == null) { "Duplicate plugin ID" }
            }
        }
    private val cancellations = mutableMapOf<RequestKey, () -> Unit>()
    private var activeSession: String? = null

    fun install(webView: WebView) {
        WebViewCompat.addWebMessageListener(
            webView,
            "__tokamakNative",
            setOf("https://$host"),
            this,
        )
    }

    fun close() {
        activeSession = null
        val cancelAll = cancellations.values.toList()
        cancellations.clear()
        cancelAll.forEach { it() }
    }

    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ) {
        plugins.values.forEach {
            it.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun onPostMessage(
        view: WebView,
        message: WebMessageCompat,
        sourceOrigin: Uri,
        isMainFrame: Boolean,
        replyProxy: JavaScriptReplyProxy,
    ) {
        if (
            !isMainFrame ||
                sourceOrigin.scheme != "https" ||
                sourceOrigin.host != host ||
                (sourceOrigin.port != -1 && sourceOrigin.port != 443)
        ) {
            return
        }
        val request = runCatching { JSONObject(message.data ?: return) }.getOrNull() ?: return
        val session = request.optString("session")
        if (session.isEmpty()) return
        if (request.optString("type") == "reset") {
            close()
            activeSession = session
            return
        }
        val id = request.optInt("id", -1)
        if (id < 0) return
        val key = RequestKey(session, id)
        when (request.optString("type")) {
            "cancel" -> cancellations.remove(key)?.invoke()
            "call", "subscribe" -> dispatch(request, key, replyProxy)
            else ->
                send(
                    replyProxy,
                    key,
                    Result.failure(
                        TokamakPluginError.notSupported("Plugin operation is not supported"),
                    ),
                    true,
                )
        }
    }

    private fun dispatch(
        request: JSONObject,
        key: RequestKey,
        replyProxy: JavaScriptReplyProxy,
    ) {
        val plugin = plugins[request.optString("plugin")]
        val method = request.optString("method")
        if (plugin == null || method.isEmpty()) {
            send(
                replyProxy,
                key,
                Result.failure(TokamakPluginError.notSupported("Plugin is not supported")),
                true,
            )
            return
        }
        val arguments = request.opt("arguments")
        val reply: TokamakPluginReply = { result ->
            send(replyProxy, key, result, request.optString("type") == "call")
        }
        if (request.optString("type") == "call") {
            plugin.call(method, arguments, reply)
        } else {
            val cancellation = plugin.subscribe(method, arguments, reply)
            cancellations[key] = cancellation
        }
    }

    private fun send(
        replyProxy: JavaScriptReplyProxy,
        key: RequestKey,
        result: Result<Any?>,
        done: Boolean,
    ) {
        if (activeSession != key.session) return
        val error = result.exceptionOrNull()
        if (done) cancellations.remove(key)?.invoke()
        val response =
            JSONObject()
                .put("session", key.session)
                .put("id", key.id)
                .put("done", done)
        if (error == null) {
            response.put("value", JSONObject.wrap(result.getOrNull()))
        } else {
            val name = (error as? TokamakPluginError)?.errorName ?: "UnknownError"
            response.put(
                "error",
                JSONObject().put("name", name).put("message", error.message ?: name),
            )
        }
        activity.runOnUiThread {
            if (activeSession == key.session) replyProxy.postMessage(response.toString())
        }
    }
}

File: platforms/apple/source/TokamakPlugins.swift
import Foundation
import WebKit

struct TokamakPluginError: Error {
  let name: String
  let message: String

  static func notSupported(_ message: String) -> Self {
    Self(name: "NotSupportedError", message: message)
  }
}

typealias TokamakPluginReply = (Result<Any?, TokamakPluginError>) -> Void

protocol TokamakPlugin: AnyObject {
  var id: String { get }

  func call(
    method: String,
    arguments: Any,
    reply: @escaping TokamakPluginReply
  )

  func subscribe(
    method: String,
    arguments: Any,
    reply: @escaping TokamakPluginReply
  ) -> (() -> Void)
}

extension TokamakPlugin {
  func call(
    method: String,
    arguments: Any,
    reply: @escaping TokamakPluginReply
  ) {
    reply(.failure(.notSupported("\(id).\(method) is not supported")))
  }

  func subscribe(
    method: String,
    arguments: Any,
    reply: @escaping TokamakPluginReply
  ) -> (() -> Void) {
    reply(.failure(.notSupported("\(id).\(method) is not supported")))
    return {}
  }
}

final class TokamakPluginBridge: NSObject, WKScriptMessageHandler {
  private struct RequestKey: Hashable {
    let session: String
    let id: Int
  }

  private let host: String
  private let plugins: [String: any TokamakPlugin]
  private var cancellations: [RequestKey: () -> Void] = [:]
  weak var webView: WKWebView?

  init(host: String, plugins: [any TokamakPlugin]) {
    self.host = host
    var pluginsById: [String: any TokamakPlugin] = [:]
    for plugin in plugins {
      precondition(pluginsById[plugin.id] == nil, "Duplicate plugin ID")
      pluginsById[plugin.id] = plugin
    }
    self.plugins = pluginsById
  }

  func install(in controller: WKUserContentController) {
    controller.add(self, name: "tokamak")
    controller.addUserScript(
      WKUserScript(
        source: Self.bootstrap(host: host),
        injectionTime: .atDocumentStart,
        forMainFrameOnly: true
      )
    )
  }

  func close() {
    let cancelAll = Array(cancellations.values)
    cancellations.removeAll()
    for cancel in cancelAll {
      cancel()
    }
  }

  func userContentController(
    _ userContentController: WKUserContentController,
    didReceive message: WKScriptMessage
  ) {
    let origin = message.frameInfo.securityOrigin
    guard
      message.frameInfo.isMainFrame,
      origin.protocol == "https",
      origin.host == host,
      origin.port == 0 || origin.port == 443,
      let encoded = message.body as? String,
      let data = encoded.data(using: .utf8),
      let request = try? JSONSerialization.jsonObject(with: data)
        as? [String: Any],
      let type = request["type"] as? String,
      let session = request["session"] as? String
    else {
      return
    }

    if type == "reset" {
      close()
      return
    }
    guard let id = request["id"] as? Int else { return }
    let key = RequestKey(session: session, id: id)
    if type == "cancel" {
      cancellations.removeValue(forKey: key)?()
      return
    }
    dispatch(type: type, key: key, request: request)
  }

  private func dispatch(
    type: String,
    key: RequestKey,
    request: [String: Any]
  ) {
    guard
      let pluginId = request["plugin"] as? String,
      let method = request["method"] as? String,
      let plugin = plugins[pluginId]
    else {
      send(
        key: key,
        result: .failure(.notSupported("Plugin is not supported")),
        done: true
      )
      return
    }
    let arguments = request["arguments"] ?? NSNull()
    let reply: TokamakPluginReply = { [weak self] result in
      self?.send(key: key, result: result, done: type == "call")
    }
    switch type {
    case "call":
      plugin.call(method: method, arguments: arguments, reply: reply)
    case "subscribe":
      let cancellation = plugin.subscribe(
        method: method,
        arguments: arguments,
        reply: reply
      )
      cancellations[key] = cancellation
    default:
      send(
        key: key,
        result: .failure(.notSupported("Plugin operation is not supported")),
        done: true
      )
    }
  }

  private func send(
    key: RequestKey,
    result: Result<Any?, TokamakPluginError>,
    done: Bool
  ) {
    if done {
      cancellations.removeValue(forKey: key)?()
    }
    var response: [String: Any] = [
      "session": key.session,
      "id": key.id,
      "done": done,
    ]
    switch result {
    case .success(let value):
      response["value"] = value ?? NSNull()
    case .failure(let error):
      response["error"] = ["name": error.name, "message": error.message]
    }
    guard
      JSONSerialization.isValidJSONObject(response),
      let data = try? JSONSerialization.data(withJSONObject: response),
      let json = String(data: data, encoding: .utf8)
    else {
      return
    }
    DispatchQueue.main.async { [weak self] in
      guard
        let self,
        self.webView?.url?.scheme == "https",
        self.webView?.url?.host == self.host,
        self.webView?.url?.port == nil || self.webView?.url?.port == 443
      else {
        return
      }
      self.webView?.evaluateJavaScript(
        "globalThis.__tokamakReceive?.(\(json))"
      )
    }
  }

  private static func bootstrap(host: String) -> String {
    """
    if (globalThis.location.origin === "https://\(host)") {
      globalThis.__tokamakNative = {
        onmessage: null,
        postMessage(message) {
          globalThis.webkit.messageHandlers.tokamak.postMessage(message);
        }
      };
    }
    """
  }
}
